Extend webhook usage section (#69)
grammyjs/website: site/docs/guide/deployment-types.md

|            |    ---  hi, new message   --->    |           |
|            | <--- okay, and sendChatAction --- |           |
|____________|                                   |___________|
```

### Ending Webhook Requests in Time

> You can ignore the rest of this page if all your middleware completes fast, i.e. within a few seconds.
> This section is for example for people who want to do file transfers or other operations that need more time.

When Telegram sends an update from one chat to your bot, it will wait for you to end the request before delivering the next update that belongs to that chat.
In other words, Telegram will deliver updates from the same chat in sequence, and updates from different chats are sent concurrently.
(The source of this information is [here](https://github.com/tdlib/telegram-bot-api/issues/75#issuecomment-755436496).)

Telegram tries to make sure that your bot receives all updates.
That means that if an update delivery fails for a chat, the subsequent updates will be queued until the first update succeeds.

#### Why Not Ending a Webhook Request Is Dangerous

Telegram has a timeout when sending an update to your webhook endpoint.

> If don't end a webhook request early enough, Telegram will resend the corresponding update, assuming it could not be delivered.

Author: Re-sending updates because of a timeout or an error are two different cases, and I'd like to keep the two things separate. Otherwise, I'd go with this:
```suggestion
If you don't end a webhook request fast enough, Telegram will re-send the update, assuming that it was not delivered.
```

|  Telegram  | --- I said new message!!! --->    |    Bot    |
|            |                              ..   |           |
|            |    *bot processing twice*    ..   |           |
|            |                              ..   |           |
|            | ---      HEEELLLOOOO      --->    |           |
|            |                              ...  |           |
|            |   *bot processing thrice*    ...  |           |
|____________|                              ...  |___________|
```

This is why grammY has its own, shorter timeout inside `webhookCallback` (default: 10 seconds).
If your middleware finishes before that, the function `webhookCallback` will respond to the webhook automatically.
In that case, everything is fine.
However, if your middleware does not finish before grammY's timeout, `webhookCallback` will throw an error.
This means that you can handle the error in your web framework.
If you don't have that error handling, Telegram will still send the same update again—but at least you have error logs now that tell you that something is wrong.

Note that that the update handling will probably not be much faster the second time the update is processed.

> As a reuslt, it will timeout once again, sending the update once again.

Reviewer: Typo in "reuslt", should be "result"

Thus, your bot will not just see the update two times, but a few dozen times, until Telegram stops retrying.
You will observe that your bot starts spamming users because it tries to handle all those updates (that are in fact the same every time).

#### Why Ending a Webhook Request Early Is Also Dangerous

You can configure `webhookCallback` to not throw an error after the timeout, but to simply end the webhook request early, even though the middleware is still running.
You can do this by passing `'return'` as a third argument to `webhookCallback`, instead of the default value `'throw'`.
However, while this behavior has its valid use cases, such a solution usually causes more problems than it solves.

Remember that once you respond to the webhook request, Telegram will send the next update for the same chat.
However, if your old update is still processing, that means that suddenly two updates are processed in parallel that were processed sequentially before.
This is going to lead to race conditions.

> For example, the session plugin will break inevitably due to WAR hazards.

Author: Good idea!

**This causes data loss!**
Other plugins and even your own middleware may break, too.
The extent of this is unknown and depends on your bot.

#### How to Solve This Problem

> This answer is very simple to say, but it can be hard to implement.

Author: Ohhh yeeeeaaah _easier said than done_ is probably what I was thinking of. Thanks!

|            |                               .   |           |
|  Telegram  |     *bot queue working*       .   |    Bot    |
|            |                               .   |           |
|            |                               .   |           |
|            |  <--- message with result  ---    |           |
|            |   ---       alrighty       --->   |           |
|____________|                                   |___________|
```

#### Why `'return'` Generally Is Worse Than `'throw'`

You may be wondering why the default action of `webhookCallback` is to throw an error, instead of ending the request successfully.
This design choice has the following reasons.

Race conditions are very hard to reproduce and may occur extremely rarely.
The solution to all of this is to _make sure not to run into timeouts_ in the first place.
But if you do, you should really want to know that this is happening, and investigate the problem!
For that reason, you want the error to occur in your logs.
Setting the timeout handler to `'return'`, hence suppressing the timeout, and pretending like nothing happened, is exactly the opposite of what should happen.

> If you do this, you're in some sense using the update queue in Telegram's webhook delivery as your task queue.

Author: I agree that it's not too clear, but I cannot come up with something better, either. I'll leave this thread open, maybe we'll find something later.

This is a bad idea for all the reasons described above.
Just because grammY _can_ suppress errors that make you lose your data, does not mean you _should_ use it.

> This configuration setting is not meant for the cases where your middleware takes too much time to complete.

Author: I like the extra sentence, good idea!
